lib/krb5/crypto-evp.c

```diff
@@ -190,8 +190,387 @@ _krb5_evp_encrypt(krb5_context context,
     return 0;
 }
 
+struct _krb5_evp_iov_cursor
+{
+    struct krb5_crypto_iov *iov;
+    int niov;
+    krb5_data current;
+    int nextidx;
+};
+
 static const unsigned char zero_ivec[EVP_MAX_BLOCK_LENGTH] = { 0 };
 
+static inline int
+_krb5_evp_iov_should_encrypt(struct krb5_crypto_iov *iov)
+{
+    return (iov->flags == KRB5_CRYPTO_TYPE_DATA
+	    || iov->flags == KRB5_CRYPTO_TYPE_HEADER
+	    || iov->flags == KRB5_CRYPTO_TYPE_PADDING);
+}
+/*
+ * If we have a group of iovecs which have been split up from
+ * a single common buffer, expand the 'current' iovec out to
+ * be as large as possible.
+ */
+
+static inline void
+_krb5_evp_iov_cursor_expand(struct _krb5_evp_iov_cursor *cursor)
+{
+    if (cursor->nextidx == cursor->niov)
+       return;
+
+    while (_krb5_evp_iov_should_encrypt(&cursor->iov[cursor->nextidx])) {
+	if ((char *)cursor->current.data + cursor->current.length
+	    != cursor->iov[cursor->nextidx].data.data) {
+            return;
+        }
+	cursor->current.length += cursor->iov[cursor->nextidx].data.length;
+	cursor->nextidx++;
+    }
+
+    return;
+}
+
+/* Move the cursor along to the start of the next block to be
+ * encrypted */
+static inline void
+_krb5_evp_iov_cursor_nextcrypt(struct _krb5_evp_iov_cursor *cursor)
+{
+    int i;
+
+    for (i = cursor->nextidx; i < cursor->niov; i++) {
```

Comment:
I think there's a termination problem here - cursor->nextidx needs to be advanced even if the last iovec isn't one that should be encrypted. Otherwise we'll never hit cursor->nextidx == niov which is required for some loop terminations.